### ISSUE TITLE
[gradle] Cannot change strategy of configuration ':backend:checkerFramework' after it has been resolved.

### ISSUE THREAD

**Issue by boris-petrov** [typetools/checker-framework#1290]
I'm trying to use Gradle with The Checker Framework. I have a main Gradle file and a subproject (called `backend`) in which I follow the guides (using JDK8):

https://checkerframework.org/manual/#gradle

I receive the following error:

```
Cannot change strategy of configuration ':backend:checkerFramework' after it has been resolved.
```

The failure is because of:

```gradle
tasks.withType(JavaCompile) {
        ....
	options.compilerArgs += configurations.checkerFramework.asPath
```

The line that the failure is on is in the main Gradle file on a:

```gradle
project.evaluationDependsOnChildren()
```

I have a couple of those in the main Gradle file. When I remove them (although I do need them so this is a temporary "fix") I still receive the same error but this time without a faulty line. Obviously something else in my configuration is "wrong".

Any ideas as to what I might try? My Gradle files are huge so pasting them here won't help too much.

**Comment by smillst** [comment]
Sorry you're having problems.  I have never seen this warning before, so I don't have any suggestions.

You said you followed the manual instructions, but the point of failure you mentioned doesn't match what the manual suggests.  So maybe that's the problem?
```
tasks.withType(JavaCompile) {
        ....
	options.compilerArgs += configurations.checkerFramework.asPath
```

```
    tasks.withType(JavaCompile).all { JavaCompile compile ->
        compile.options.compilerArgs = [
                '-processor', 'org.checkerframework.checker.nullness.NullnessChecker',
                '-processorpath', "${configurations.checkerFramework.asPath}",
                // uncomment to turn Checker Framework errors into warnings
                // '-Awarns',
                "-Xbootclasspath/p:${configurations.checkerFrameworkAnnotatedJDK.asPath}"
        ]
```

**Comment by boris-petrov** [comment]
Thanks for the response! Actually I did try following the exact instructions but the effect was the same (besides, the code that I pasted and the one in the manual are pretty much the same).

Anyway, I still have no idea why this happens. If anyone has a solution, I'll be glad to hear it. Otherwise, let's keep this open and I'll notify you of any progress. Thanks!

**Comment by mernst** [comment]
Without a concrete test case, we won't be able to understand, fix, or test this.
You remarked that your gradle files are huge.  Could you create a smaller (preferably minimal!) example?
Thanks.

**Comment by boris-petrov** [comment]
I will try to do that next week at some point. I hope I can help with the reproduction.

**Comment by mernst** [comment]
Were you able to create a smaller example that reproduces the problem?

**Comment by boris-petrov** [comment]
Yes, sorry for the delay, got busy with other things. So in the end the problem is either with the `Checker Framework` or with [Gretty](https://github.com/akhikhl/gretty) - but these two things do not work together. Here as an example `Gradle` script:

```gradle
plugins {
	id 'org.akhikhl.gretty' version '1.4.2'
}

apply plugin: 'java'
apply plugin: 'org.akhikhl.gretty'

repositories {
	mavenCentral()
}

def checkerFrameworkVersion = '2.1.12'

configurations {
	checkerFrameworkAnnotatedJDK {
		description = 'A copy of JDK classes with Checker Framework type qualifers inserted'
	}

	checkerFramework {
		description = 'The Checker Framework: custom pluggable types for Java'
	}
}

dependencies {
	checkerFrameworkAnnotatedJDK "org.checkerframework:jdk8:${checkerFrameworkVersion}"
	checkerFramework "org.checkerframework:checker:${checkerFrameworkVersion}"
	compile "org.checkerframework:checker-qual:${checkerFrameworkVersion}"
}

tasks.withType(JavaCompile) {
	options.compilerArgs += '-processor'
	options.compilerArgs += 'org.checkerframework.checker.nullness.NullnessChecker'
	options.compilerArgs += '-processorpath'
	options.compilerArgs += configurations.checkerFramework.asPath
	options.compilerArgs += '-Awarns'
	options.compilerArgs += "-Xbootclasspath/p:${configurations.checkerFrameworkAnnotatedJDK.asPath}"
}

task foo
```

If you run a `gradle foo` on this, it will blow up with the error I've specified. If you remove lines 2 and 6 (the `Gretty` parts) - it will work fine.

Is this something you can fix or should I ask the `Gretty` people? `Gretty` is a pretty popular thing so I'm really surprised that no one has hit this before me. Unfortunately it's not maintained too much these days and they probably won't look at it any time soon.

Please tell me if you need any more information. Thanks a lot for the support!

**Comment by mernst** [comment]
Thanks.  I can reproduce the problem, running either `gradle build` or `gradle foo`, with either Gradle 3.1 or 3.5.

The error I get is slightly different than yours:

```
FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring root project 'cf-gradle-bug'.
> Cannot change strategy of configuration ':checkerFramework' after it has been resolved.
```

but surely has the same cause

A Google search for
    ```gretty Cannot change strategy of configuration after it has been resolved.```
or
    ```gradle "Cannot change strategy of configuration" "after it has been resolved."```
yields quite a few reports of the same problem, and some suggestions for fixing the problem.

We aren't experts on the intricacies of every build system, but will try to look into this.  Any suggestions from other people (exspecially gradle experts) are welcome.

**Comment by boris-petrov** [comment]
Thank you for the support!

Nothing I did as a fix worked - and in the small example I gave you, unfortunately there isn't much to be tried. :( One fix for a similar problem in `gradle-apt-plugin` is this (according to [this](https://github.com/tbroyer/gradle-apt-plugin/issues/31) issue):

https://github.com/tbroyer/gradle-apt-plugin/commit/6c137f71f069b621ed7834314e59327bde289495

I'm not sure if that helps in any way though. I really hope you can resolve this as I want to try the `Checker Framework`.

**Comment by boris-petrov** [comment]
`Gretty` 2.0.0 was released a few days ago and this issue is no longer. I'm closing this, thanks for the support! However, there is a new issue if you'd like to help. :)

https://github.com/typetools/checker-framework/issues/1362
